pkg/function/service.go

```
@@ -41,9 +84,65 @@ func (fs *Service) DoFn(ctx context.Context, d *functionpb.Datum) (*functionpb.D
 }
 
 // ReduceFn applies a reduce function to a datum stream.
-// TODO: implement ReduceFn
-func (fs *Service) ReduceFn(fnServer functionpb.UserDefinedFunction_ReduceFnServer) error {
-	ctx := fnServer.Context()
-	_ = ctx
-	return nil
+func (fs *Service) ReduceFn(stream functionpb.UserDefinedFunction_ReduceFnServer) error {
+	var (
+		ctx      = stream.Context()
+		key      string
+		reduceCh = make(chan Datum)
+		md       Metadata
+	)
+
+	// get key and metadata from gPRC metadata
+	if grpcMD, ok := metadata.FromIncomingContext(ctx); ok {
+		// get Key
+		keyValue := grpcMD.Get(DatumKey)
+		if len(keyValue) > 1 {
+			return fmt.Errorf("expect exact one key but got multiple keys")
+		} else if len(keyValue) == 0 {
+			return fmt.Errorf("missing key")
+		}
+		key = keyValue[0]
+		// TODO: get metadata
+	}
+
+	var (
+		datumList []*functionpb.Datum
+		wg        sync.WaitGroup
+	)
+
+	go func() {
+		wg.Add(1)
+		defer wg.Done()
+		messages, err := fs.Reducer.HandleDo(ctx, key, reduceCh, md)
+		if err != nil {
+			// TODO: deal with err
+			// will return an empty datumList
+			return
+		}
+		for _, msg := range messages {
+			datumList = append(datumList, &functionpb.Datum{
+				Key:       msg.Key,
+				Value:     msg.Value,
+				EventTime: &functionpb.EventTime{EventTime: timestamppb.New(time.Time{})}, // TODO: what's the correct value?...
+				Watermark: &functionpb.Watermark{Watermark: timestamppb.New(time.Time{})},
+			})
+		}
+	}()
+
+	for {
+		datum, err := stream.Recv()
+		if err == io.EOF {
```

we are swallowing other errors here